sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -218,7 +247,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       bucketSpec = getBucketSpec,
       options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    val destination = source match {
+      case "jdbc" => extraOptions.get(JDBCOptions.JDBC_TABLE_NAME)
+      case _ => extraOptions.get("path")
```

Comment:
As we already use `Map` as the user-facing API for `DataFrameWriter`, I think we can't change it. To make it consistent, I think it's reasonable to still pass `Map` to listeners. I agree it's bad that listeners to know these magic keys, but it's already the case of `DataFrameWriter`/